File: 251051/Korneliuk/NoSQL/publisher/src/main/java/com/rest/restapp/client/DiscussionClient.java
package com.rest.restapp.client;

import com.rest.restapp.dto.request.NoteRequestToDto;
import com.rest.restapp.dto.response.NoteResponseTo;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;

import java.util.List;

@Component
public class DiscussionClient {

    private final WebClient webClient;

    public DiscussionClient() {
        this.webClient = WebClient.builder()
                .baseUrl("http://localhost:24130/api/v1.0")
                .build();
    }

    public NoteResponseTo createNote(NoteRequestToDto request) {
        return webClient.post()
                .uri("/notes")
                .bodyValue(request)
                .retrieve()
                .bodyToMono(NoteResponseTo.class)
                .block();
    }

    public List<NoteResponseTo> getAll() {
        return webClient.get()
                .uri("/notes")
                .retrieve()
                .bodyToFlux(NoteResponseTo.class)
                .collectList()
                .block();
    }

    public NoteResponseTo getById(Long id) {
        return webClient.get()
                .uri("/notes/{id}", id)
                .retrieve()
                .bodyToMono(NoteResponseTo.class)
                .block();
    }

    public NoteResponseTo update(Long id, NoteRequestToDto request) {
        return webClient.put()
                .uri("/notes/{id}", id)
                .bodyValue(request)
                .retrieve()
                .bodyToMono(NoteResponseTo.class)
                .block();
    }

    public List<NoteResponseTo> getNotesByIssueId(Long issueId) {
        return webClient.get()
                .uri("/notes/issue/" + issueId)
                .retrieve()
                .bodyToFlux(NoteResponseTo.class)
                .collectList()
                .block();
    }
    
    public ResponseEntity<Void> deleteNote(Long noteId) {
        return webClient.delete()
                .uri("/notes/{id}", noteId)
                .retrieve()
                .toBodilessEntity()
                .block();
    }
}

File: 351004/Barsukov/app/clients/discussion_client.py
import httpx
from typing import List, Optional
from schemas.note import NoteRequestTo, NoteResponseTo


class DiscussionClient:
    def __init__(self):
        self.base_url = "http://localhost:24130/api/v1.0"

    async def create_note(self, dto: NoteRequestTo) -> Optional[NoteResponseTo]:
        async with httpx.AsyncClient() as client:
            response = await client.post(f"{self.base_url}/notes", json=dto.model_dump())
            if response.status_code == 201:
                return NoteResponseTo(**response.json())
            return None

    async def get_notes_by_issue(self, issue_id: int) -> List[NoteResponseTo]:
        async with httpx.AsyncClient() as client:
            response = await client.get(f"{self.base_url}/issues/{issue_id}/notes")
            if response.status_code == 200:
                return [NoteResponseTo(**item) for item in response.json()]
            return []

    async def get_note(self, issue_id: int, note_id: int) -> Optional[NoteResponseTo]:
        async with httpx.AsyncClient() as client:
            response = await client.get(f"{self.base_url}/issues/{issue_id}/notes/{note_id}")
            if response.status_code == 200:
                return NoteResponseTo(**response.json())
            return None

    async def update_note(self, issue_id: int, note_id: int, dto: NoteRequestTo) -> Optional[NoteResponseTo]:
        async with httpx.AsyncClient() as client:
            response = await client.put(
                f"{self.base_url}/issues/{issue_id}/notes/{note_id}",
                json=dto.model_dump()
            )
            if response.status_code == 200:
                return NoteResponseTo(**response.json())
            return None

    async def delete_note(self, issue_id: int, note_id: int) -> bool:
        async with httpx.AsyncClient() as client:
            response = await client.delete(f"{self.base_url}/issues/{issue_id}/notes/{note_id}")
            return response.status_code == 204

    async def update_note_state(self, issue_id: int, note_id: int, state: str) -> bool:
        async with httpx.AsyncClient() as client:
            response = await client.patch(
                f"{self.base_url}/issues/{issue_id}/notes/{note_id}/state",
                json={"state": state}
            )
            return response.status_code == 200